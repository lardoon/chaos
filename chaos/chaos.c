#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include "chaos/platform.h"
#include "chaos/porting.h"
#include "chaos/chaos.h"
#include "chaos/options.h"
#include "chaos/text16.h"
#include "chaos/rand.h"

const unsigned short int chaos_cols[] = {
	RGB16(31, 31, 31),	/* white */
	RGB16(31, 31, 0),	/* yellow */
	RGB16(0, 31, 31),	/*lightblue */
	RGB16(0, 31, 0),	/*green */
	RGB16(31, 0, 31),	/*purple */
	RGB16(31, 0, 0),	/*red */
	RGB16(0, 0, 31),	/*blue */
	RGB16(20, 20, 20),	/*grey */
	RGB16(20, 20, 0),	/* mustard */
};

struct chaos_state g_chaos_state;

/* on the player select screen, the colours are:
   red purple green lightblue mustard yellow grey white
*/
const signed char surround_table[8][2] = {
	{-1, -1},
	{-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}
};

const unsigned char key_table[8] = {
	'Q', 'W', 'E', 'D', 'C', 'X', 'Z', 'A'
};
int hilite_item;
signed char anim_col;
signed char anim_col_grad;

void play_soundfx(int soundfx)
{
	if (Options[OPT_SOUND_ENABLED])
		platform_play_soundfx(soundfx);
}

void append_stat(char *buffer, const char *name, int value)
{
	char tmp[10];
	strcat(buffer, "\t");
	strcat(buffer, name);
	strcat(buffer, " ");
	int2a(value, tmp, 10);
	strcat(buffer, tmp);
	strcat(buffer, "\n");
}

static void append_seed(char *buffer)
{
	/* SEED 0xFFFFAAAA */
	static const char *hex = "0123456789ABCDEF";
	strcat(buffer, "\tSEED 0x");
	unsigned int seed = GetSeed();
	/* append as hex... */
	int i;
	size_t pos = strlen(buffer);
	for (i = 28 ; i >= 0; i -= 4) {
		buffer[pos++] = hex[(seed >> i) & 0xf];
	}
	buffer[pos++] = '\n';
	buffer[pos++] = 0;
}

void chaos_state_to_str(char *buffer)
{
	strcat(buffer, "STATE:\n");
	append_stat(buffer, "CHAOS", g_chaos_state.world_chaos);
	append_stat(buffer, "ROUNDS", g_chaos_state.round_count);
	append_stat(buffer, "KILLED", g_chaos_state.dead_wizards);
	append_stat(buffer, "PLAYERS", g_chaos_state.playercount);
	append_stat(buffer, "SCREEN", g_chaos_state.current_screen);
	append_stat(buffer, "CURRENT", g_chaos_state.current_player);
	append_seed(buffer);
}

static void parse_line(const char *line)
{
	const char *tmp = line;
	while (isspace((unsigned char)*tmp)) {
		tmp++;
	}
	const char *namestart = tmp;
	while (!isspace((unsigned char)*tmp))
		tmp++;
	char *nameend = (char *)tmp;
	while (isspace((unsigned char)*tmp)) {
		tmp++;
	}
	const char *valstart = tmp;
	while (*tmp != '\n')
		tmp++;
	char *valend = (char *)tmp;
	char tmpvalend = *valend;
	char tmpnameend = *nameend;
	*valend = 0;
	*nameend = 0;
	if (strcmp(namestart, "KILLED") == 0) {
		g_chaos_state.dead_wizards = strtol(valstart, 0, 0);
	} else if (strcmp(namestart, "CHAOS") == 0) {
		g_chaos_state.world_chaos = strtol(valstart, 0, 0);
	} else if (strcmp(namestart, "ROUNDS") == 0) {
		g_chaos_state.round_count = strtol(valstart, 0, 0);
	} else if (strcmp(namestart, "PLAYERS") == 0) {
		g_chaos_state.playercount = strtol(valstart, 0, 0);
	} else if (strcmp(namestart, "SCREEN") == 0) {
		g_chaos_state.current_screen = strtol(valstart, 0, 0);
	} else if (strcmp(namestart, "CURRENT") == 0) {
		g_chaos_state.current_player = strtol(valstart, 0, 0);
	} else if (strcmp(namestart, "SEED") == 0) {
		unsigned int seed = strtoul(valstart, 0, 0);
		setSeed(seed);
	}
	*valend = tmpvalend;
	*nameend = tmpnameend;
}

int get_save_version(const char *data)
{
	if (strncmp(data, "VERSION", 7) != 0) {
		return 0;
	}
	const char *tmp = &data[7];
	if (*tmp == 0) {
		return 0;
	}
	while (isspace((unsigned char)*tmp)) {
		tmp++;
	}
	char num[10];
	int i;
	for (i = 0; i < 9 && isdigit((unsigned char)*tmp); i++, tmp++) {
		num[i] = *tmp;
	}
	num[i] = 0;
	return strtol(num, 0, 0);
}

void set_save_version(char *buffer, int version)
{
	char tmp[10];
	strcat(buffer, "VERSION ");
	int2a(version, tmp, 10);
	strcat(buffer, tmp);
	strcat(buffer, "\n");
}

void str_to_chaos_state(char *data)
{
	if (data == NULL)
		return;
	const char *tmp = data;
	int in_state = 0;
	while (*tmp != 0) {
		const char *linebegin = tmp;
		while (*tmp != '\n' && *tmp != 0) {
			tmp++;
		}
		tmp++;
		if (linebegin[0] != '\t') {
			if (strncmp(linebegin, "STATE:\n", 7) == 0)
				in_state = 1;
			else if (in_state != 0)
				break;
		} else {
			parse_line(linebegin);
		}
	}
}
