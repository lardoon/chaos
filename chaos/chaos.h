#ifndef chaos_h_seen
#define chaos_h_seen
/* chaos.h */

#include <stdint.h>

#define CHAOS_SAVE_VERSION       1

#define SCR_SPLASH               10
#define SCR_OPTIONS              20
#define SCR_CREATE_PLAYERS       30
#define SCR_EDIT_NAME            40
#define SCR_GAME_MENU            50
#define SCR_SELECT_SPELL         60
#define SCR_EXAMINE_SPELL        70
#define SCR_EXAMINE_BOARD        80

#define SCR_CASTING              100
#define SCR_MOVEMENT             110
#define SCR_INFO_SCREEN          120
#define SCR_EXAMINE              130

/* should be last arena spell + 1 */
#define WIZARD_INDEX              43

/* defines for status_flag */
#define UNDEAD 1
#define ENGAGED 2
#define MOVELEFT 4
#define TURNOVER 8
#define ISHUMAN 16
#define SPELL_IS_ILLUSION 32
#define ISHIDDEN 64
#define ALIVE 128
struct chaos_state {
	signed char world_chaos;	/* Chaos level in the world */
	unsigned char round_count;	/* total rounds played so far */
	unsigned char dead_wizards;	/* number of dead wizards */
	unsigned char playercount;	/* number of total players 2-8 */
	unsigned char current_player;	/* current player to move/cast */
	unsigned long int current_screen;
};

extern struct chaos_state g_chaos_state;

extern const unsigned short int chaos_cols[];
extern int hilite_item;
extern signed char anim_col;
extern signed char anim_col_grad;
extern volatile uint8_t game_frames;
extern const signed char surround_table[8][2];
extern const unsigned char key_table[8];
int enable_interrupts(void);
int disable_interrupts(void);
void delay(int d);

void play_soundfx(int);

/* appends a stat to the buffer with the given name */
void append_stat(char *buffer, const char *name, int value);

/** convert g_chaos_state to a string.*/
void chaos_state_to_str(char *buffer);
/** given a string, fill out g_chaos_state. */
void str_to_chaos_state(char *data);

/** From the save data, get save version. This is the first thing in the file. */
int get_save_version(const char *data);
/** Add the VERSION = version string to the buffer. @see CHAOS_SAVE_VERSION. */
void set_save_version(char *buffer, int version);

#endif
