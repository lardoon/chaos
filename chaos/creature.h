/* creature.h */
#ifndef CREATURE_H
#define CREATURE_H
/* bit 6 is undead flag */
#define IS_UNDEAD(n) ((n & BIT06) == BIT06)
/* bit 4 is illusion */
#define IS_ILLUSION(n) ((n & BIT04) == BIT04)
/* bit 4 is asleep */
#define IS_ASLEEP(n) ((n & BIT03) == BIT03)

#include "chaos/spellenums.h"
#define IS_FLYING_MOUNT(c) ( (c) >= SPELL_PEGASUS && (c) <= SPELL_MANTICORE)

void creature_spell_succeeds(unsigned char target);
int draw_all_creatures(void);
void spread_fire_blob(void);
void clear_arena(void);
void liberate_from_blob(int idx);
/**
 * Invalidate the drawn-creature cache
 */
void invalidate_cache(void);

#endif				/* CREATURE_H */
