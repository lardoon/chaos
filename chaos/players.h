/* players.h */

#ifndef PLAYERS_H
#define PLAYERS_H
int show_create_players(void);
int redraw_create_players(void);
void create_players_up(void);
void create_players_down(void);
void create_players_left(void);
void create_players_right(void);
void create_players_l(void);
void create_players_r(void);
void create_players_accept(void);
void create_players_start(void);
void create_players_touch(int x, int y);
void create_players_back(void);
void animate_player_screen(void);
unsigned char get_next_human(unsigned char id);
void new_random_spell(intptr_t arg, int index);
void init_players(void);
void create_default_wizs(void);

void players_to_str(char *buffer);
void parse_players(char *plystr);

#endif				/* PLAYERS_H */

