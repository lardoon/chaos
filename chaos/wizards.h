/* wizards.h */

/* the wizard definitions */
/* the ser colour to change.. */
#ifndef WIZARDS_H
#define WIZARDS_H

#define WIZARD_COLOUR 0x1

/* shadow from is bit 3 */
#define HAS_SHADOWFORM(flag) ( (flag & 0x8) == 0x8 )
/* if bit 4 set, wizard is dead */
#define IS_WIZARD_DEAD(flag) ( (flag & 0x10) == 0x10 )
/* wings is bit 5 */
#define HAS_MAGICWINGS(flag) ( (flag & 0x20) == 0x20 )

#define GFX_MAGIC_SHIELD  8
#define GFX_MAGIC_ARMOUR  9
#define GFX_MAGIC_SWORD   10
#define GFX_MAGIC_KNIFE   11
#define GFX_MAGIC_BOW     12
#define GFX_MAGIC_WINGS   13

/* human and cpu are obvious */
#define PLYR_HUMAN        0
#define PLYR_CPU          1

#define IS_CPU(n) ((players[n].plyr_type&PLYR_CPU)>0)
extern unsigned char current_player;
typedef struct  {
	char name[12];
	unsigned char combat;
	unsigned char ranged_combat;
	unsigned char range;
	unsigned char defence;
	unsigned char movement_allowance;
	unsigned char manoeuvre_rating;
	unsigned char magic_resistance;
	unsigned char spell_count;
	unsigned char spells[42];	/* 2 byte pairs, one for the priority, the other for the spell id */
	unsigned char ability;

	/*animationspeed */
	unsigned char image;	/* this is different... */
	unsigned short colour;

	unsigned char plyr_type;
	unsigned char modifier_flag;
	unsigned char illusion_cast;
	unsigned char selected_spell;
	unsigned char last_spell;
	unsigned char timid;
	unsigned char team;
} player_data;		/* 60 bytes * 8 = 480 bytes of wizard data */
extern player_data *players;

/*
   meaning of bits in modifier flag...
   bit 0
   bit 1    magic knife
   bit 2    magic sword
   bit 3    shadow form
   bit 4    is dead
   bit 5    magic wings
   bit 6    magic shield
   bit 7    magic armour

*/

struct WIZARD_DATA  {
	const unsigned short *pGFX;
	const unsigned short *pMap;
};
extern const struct WIZARD_DATA WizardGFX[];
int creature_count(int player_id, int include_blob);
int update_creaturecount(void);

/* kills the wizard at the target_index */
/* updates flags, etc */
void kill_wizard(void);
void reset_players(void);
void init_player_table(void);

#endif				/* WIZARDS_H */

