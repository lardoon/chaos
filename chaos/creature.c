/* creature.c */
#include <string.h>
#include "chaos/platform.h"
#include "chaos/porting.h"
#include "chaos/sound_data.h"
#include "chaos/spelldata.h"
#include "chaos/creature.h"
#include "chaos/arena.h"
#include "chaos/chaos.h"
#include "chaos/wizards.h"
#include "chaos/casting.h"
#include "chaos/gfx.h"
#include "chaos/options.h"
#include "chaos/rand.h"

static void uncover_square(unsigned char start, unsigned char target)
{
	unsigned char r = GetRand(10);
	if (r > 2) {
		return;
	}
	unsigned char target_creature = arena[0][target];

	if (target_creature != SPELL_MAGIC_FIRE) {
		r = GetRand(10);
		if (r > 3) {
			return;
		}
	}
	/* remove the blob and show the creature underneath... */
	if (arena[4][start] == 0) {
		arena[0][start] = 0;
		arena[5][start] = 0;
	} else {
		liberate_from_blob(start);
	}
	delay(24);
}

/* code from a09b */
static void do_spread(unsigned char start, unsigned char target)
{
	unsigned char start_creature = arena[0][start];

	arena[0][target] = start_creature;
	arena[2][target] = 0;	/* frame */
	arena[3][target] = arena[3][start];	/* set owner */

	arena[3][target] |= 0x80;	/* set spread flag */
	arena[3][start] |= 0x80;	/* set spread flag */

	move_screen_to(target);
	if (start_creature == SPELL_GOOEY_BLOB) {
		play_soundfx(SND_GOOEY);
	} else if (start_creature == SPELL_MAGIC_FIRE) {
		play_soundfx(SND_FIRE);
	}

	delay(24);

	/* at the end of everything, randomly uncover the "spawn" square */
	uncover_square(start, start);
}

void creature_spell_succeeds(unsigned char target)
{
	if (!(current_spell == SPELL_GOOEY_BLOB
	      || current_spell == SPELL_MAGIC_FIRE)) {
		arena[5][target] = arena[0][target];	/* whatever is in this square, place in arena 5... */
	}
	arena[0][target] = current_spell;
	arena[1][target] = 1;
	arena[2][target] = 0;
	arena[3][target] = g_chaos_state.current_player;
	if (temp_illusion_flag) {
		arena[3][target] |= 0x10;	/* set bit 4 */
	}
}

static struct arena_cache {
	uint8_t creature;
	uint8_t frame;
} s_cache[ARENA_SIZE];

void invalidate_cache(void)
{
	memset(s_cache, 0, sizeof(s_cache));
}

void clear_arena(void)
{
	int x, y;
	for (x = 0; x < 15; x++) {
		for (y = 0; y < 10; y++) {
			clear_square(x, y);
		}
	}
	invalidate_cache();
}

/* this should be called every now and again... */
int draw_all_creatures(void)
{
	int i;
	uint8_t x, y;
	for (i = 0; i < ARENA_SIZE; i++) {
		get_yx(i, &y, &x);
		if (x == 17)
			continue;
		if (arena[0][i] >= SPELL_KING_COBRA) {
			if (s_cache[i].creature == arena[0][i] &&
					s_cache[i].frame == arena[2][i])
				continue;
			if (arena[0][i] < WIZARD_INDEX) {
				draw_creature(x - 1, y - 1,
						arena[0][i], arena[2][i]);
			} else {
				int wizard = arena[0][i] - WIZARD_INDEX;
				if (wizard < 8) {
					draw_wizard(x - 1, y - 1,
						    players[wizard].image,
						    arena[2][i],
						    wizard);
				}
			}
			s_cache[i].creature = arena[0][i];
			s_cache[i].frame = arena[2][i];
		} else {
			if (s_cache[i].creature == 0xff)
				continue;
			if (x > 0 && x < 0x10 && y > 0 && y < 0xB)
				clear_square(x - 1, y - 1);
			s_cache[i].creature = 0xff;
		}

	}
	return 0;
}

/* called at the end of the casting round */
/* based on code at 9f50..  */
void spread_fire_blob(void)
{
	int i = 0, current_index = 0;
	unsigned char this_owner, target_owner, r, creature, target_creature;
	unsigned char attack, defence;
	current_index = 0;
	target_index = 0;
	unset_moved_flags();
	for (i = 0; i < 0x9f; i++) {
		target_index = current_index;

		if (HAS_MOVED(arena[3][current_index])) {
			current_index++;
			continue;
		}
		creature = arena[0][target_index];
		if (creature < SPELL_GOOEY_BLOB) {
			current_index++;
			continue;
		}
		if (creature >= SPELL_MAGIC_WOOD) {
			current_index++;
			continue;
		}
		/* sleeping blob does not spread */
		if (is_asleep(i)) {
			arena[3][i] |= 0x80;
			continue;
		}
		this_owner = get_owner(target_index);
		r = GetRand(10);

		if (r < 0x9) {
			if (creature != SPELL_MAGIC_FIRE) {
				r = GetRand(10);
				if (r > 0x8) {
					/* jump to a0c6... */
					/* do uncover creature routine */
					uncover_square(current_index,
							target_index);
					current_index++;
					continue;
				}
			}
			r = GetRand(10);
			while (r >= 0x8) {
				r = GetRand(10);
			}
			target_index =
				apply_position_modifier(target_index, r);
			if (target_index == 0) {
				current_index++;
				continue;
			} else {
				target_index--;
			}

			/* do in range check...  */
			uint16_t dist;
			get_distance(current_index, target_index, &dist);
			if (dist > 4) {
				current_index++;
				continue;
			}

			target_creature = arena[0][target_index];
			if (target_creature != 0) {
				if (target_creature < WIZARD_INDEX) {
					/* is NOT a wizard... */
					if (target_creature >= SPELL_MAGIC_CASTLE
						|| target_creature == SPELL_MAGIC_WOOD) {
						/* a magic wood or castle, etc */
						/* jump to a0c6... */
						/* do uncover creature routine */
						uncover_square(current_index, target_index);
						current_index++;
						continue;
					}

					if (!is_dead(target_index)) {
						/* if not dead */
						target_owner = get_owner(target_index);
						if (target_owner == this_owner) {
							current_index++;
							continue;
						}
						if (target_creature <= SPELL_MANTICORE
								&& target_creature >= SPELL_HORSE
								&& arena[4][target_index] >= WIZARD_INDEX) {
							/* if the creature is a mount and has a wizard on it */
							/* a011... */
							if ((arena[4][target_index] - WIZARD_INDEX) == this_owner) {
								/* do nothing - own blob can't kill */
								current_index++;
								continue;
							}
							/* clear arena 4 (wizard) and place wizard in arena 0... */
							arena[0][target_index] = arena[4][target_index];
							arena[4][target_index] = 0x00;

							/* a033... kill wizard */
							kill_wizard();
							remove_cursor();
							/* jump to a09b  */
							do_spread(current_index, target_index);
							current_index++;
							continue;

						}
						/* a03d */
						if (creature == SPELL_MAGIC_FIRE) {
							/* jump a075 - for fire */
							/* get defence of target creature */
							defence = CHAOS_SPELLS[target_creature].defence + GetRand(10);
							attack = GetRand(10) + 5;
							if (attack > defence) {
								/* jump a0c6 */
								uncover_square(current_index, target_index);
								current_index++;
								continue;
							}
							/* jump a09b... */
						} else {	/*if (creature == SPELL_GOOEY_BLOB)*/
							/* gooey blob */
							if (target_creature >= SPELL_MAGIC_FIRE) {
								/* jump a0c6 - uncover creature routine */
								uncover_square(current_index, target_index);
								current_index++;
								continue;
							}
							if (!is_dead(target_index)) {
								/* creature alive.. set arena 4 with the current creature and cover it */
								arena[4][target_index] = arena[0][target_index];
								if (Options[OPT_OLD_BUGS]) {
									/* old bugs on... */
									/* a066 */
									/* I think the plan here was to set the creature's owner, the first 3 bits */
									arena[5][target_index] = get_owner(target_index);	/* bug here */
								} else {
									/* old bug removed here... */
									/* what about undead and illusionary flags? */
									arena[5][target_index] = arena[3][target_index];
								}
								/* spread blindness and sleep */
								if (is_asleep(target_index)) {
									arena[5][target_index] |= 8;
									arena[5][target_index] |= 0x80;
								}
								if (is_blind(target_index))
									arena[5][target_index] |= 8;
							}	/* else jump to a09b - target dead */
						}

					}	/* else target is dead */
						/* jump to a09b */
				} else {
					/* "jump a011" - target creature is a wizard */
					if ((target_creature - WIZARD_INDEX) == this_owner) {
						/* do nothing - own blob can't kill */
						current_index++;
						continue;
					}
					/* a033 - kill wizard */
					kill_wizard();
					remove_cursor();
					/* jmp a09b */
				}
			}	/* else  target creature == 0 */
			/* jump a09b */
			do_spread(current_index, target_index);
		} else {
			/* jump to a0c6... */
			/* do uncover creature routine */
			uncover_square(current_index, target_index);
		}
		current_index++;
	}
}

void liberate_from_blob(int start)
{
	arena[0][start] = arena[4][start];
	arena[4][start] = 0;
	arena[3][start] = arena[5][start];
	arena[5][start] = 0;
	int bits = arena[3][start];
	/* blind bit set, moved set */
	if ((bits & 0x88) == 0x88) {
		/* take advantage of move flag to indicate sleep */
		arena[2][start] = 4;
	}
}

