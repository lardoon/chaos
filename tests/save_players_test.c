#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>
#include "chaos/wizards.h"
#include "chaos/chaos.h"
#include "port/linux/utils.h"
#include "chaos/creature.h"
#include "chaos/players.h"

#define EXPECTED_TXT SRCDIR "/save_players_test_expected.txt"
void parse_players(char *plystr);
void create_default_wizs(void);

static void check_equals(player_data *expected, int id)
{
	player_data *p = &players[id];
	int i;
	int namediff = strcmp(p->name, expected->name);
	if (namediff)
		printf("%s != %s\n", p->name, expected->name);
	assert(namediff == 0);
	assert(expected->combat == p->combat);
	assert(expected->ranged_combat == p->ranged_combat);
	assert(expected->range == p->range);
	assert(expected->defence == p->defence);
	assert(expected->movement_allowance == p->movement_allowance);
	assert(expected->manoeuvre_rating == p->manoeuvre_rating);
	assert(expected->magic_resistance == p->magic_resistance);
	assert(expected->spell_count == p->spell_count);
	for (i = 0; i < p->spell_count; i++) {
		int idx = i * 2;
		assert(expected->spells[idx] == p->spells[idx]);
		assert(expected->spells[idx + 1] == p->spells[idx + 1]);
	}
	assert(expected->ability == p->ability);
	assert(expected->image == p->image);
	assert(expected->colour == p->colour);
	assert(expected->plyr_type == p->plyr_type);
	assert(expected->modifier_flag == p->modifier_flag);
	assert(expected->illusion_cast == p->illusion_cast);
	assert(expected->selected_spell == p->selected_spell);
	assert(expected->last_spell == p->last_spell);
	assert(expected->timid == p->timid);
}

void test_save_players(void)
{
	init_player_table();
	init_arena_table();
	setSeed(192);
	g_chaos_state.playercount = 8;
	create_default_wizs();
	init_players();

	char *result = malloc(500 * 8);
	players_to_str(result);
	char *data = read_file_as_str(EXPECTED_TXT);
	assert(data != 0);
	int diff = strcmp(data, result);
	if (diff) {
		fprintf(stderr, "%s\n", result);
	}
	assert(diff == 0);

	/* scramble players */
	setSeed(111);
	create_default_wizs();
	init_players();
	/* make sure load works */
	parse_players(result);

	/* make sure parse_players didn't screw up the data */
	diff = strcmp(data, result);
	assert(diff == 0);
	{
		player_data tmp = {
			"Rich Q",
			1, 0, 0, 1, 1, 4, 10, 13,
			{0,1,23,41,12,42,18,26,18,30,18,3,11,
				50,18,23,18,45,18,17,18,23,18,37,18,13},
			0, 5, 31775, 0, 0, 0, 0, 0, 0,
		};
		check_equals(&tmp, 0);
	}
	{
		player_data tmp = {
			"Ben",
			7, 0, 0, 4, 1, 7, 6, 15,
			{21,1,18,20,18,5,18,12,18,4,18,9,18,23,18,43,18,
				29,18,17,18,26,14,56,6,58,6,60,5,59,},
			0, 4, 32736, 65, 0, 0, 0, 0, 0,
		};
		check_equals(&tmp, 1);
	}
	{
		player_data tmp = {
			"Saruman",
			6, 0, 0, 6, 1, 7, 9, 15,
			{18,10,18,35,18,64,18,17,18,26,18,20,18,11,18,
				6,18,9,18,18,18,61,13,1,11,49,6,58,5,59},
			0, 3, 21140, 65, 0, 0, 0, 0, 0,
		};
		check_equals(&tmp, 2);
	}
	{
		player_data tmp = {
			"Kiff",
			7, 0, 0, 7, 1, 5, 10, 15,
			{18,17,18,29,18,6,18,45,18,4,18,22,18,6,18,16,18,21,18,33,18,26,18,36,17,55,14,1,6,58,},
			1, 6, 660, 65, 0, 0, 0, 0, 0,
		};
		check_equals(&tmp, 3);
	}
	{
		player_data tmp = {
			"Grotbags",
			5, 0, 0, 4, 1, 6, 9, 15,
			{23,41,18,39,18,14,18,35,18,17,18,33,18,15,18,33,18,6,18,9,18,37,18,3,18,5,18,11,15,1,},
			0, 2, 992, 65, 0, 0, 0, 0, 0,
		};
		check_equals(&tmp, 4);
	}
	{
		player_data tmp = {
			"Tim",
			7, 0, 0, 4, 1, 6, 6, 18,
			{ 18,1,18,28,18,15,18,31,18,3,18,35,18,36,18,6,18,30,18,33,18,19,18,33,18,36,18,23,12,48,12,42,12,42,11,50,},
			0, 0, 31744, 65, 0, 0, 0, 0, 0,
		};
		check_equals(&tmp, 5);
	}
	{
		player_data tmp = {
			"Sooty",
			4, 0, 0, 5, 1, 5, 6, 16,
			{18,22,18,20,18,14,18,17,18,35,18,5,18,32,18,39,18,32,18,31,18,61,18,15,18,32,18,10,18,45,14,1,},
			0, 1, 31, 65, 0, 0, 0, 0, 0,
		};
		check_equals(&tmp, 6);
	}
	{
		player_data tmp = {
			"H.Potter",
			6,
			0,
			0,
			3,
			1,
			6,
			6,
			17,
			{23,40,23,52,21,1,18,62,18,17,18,29,18,34,18,5,18,32,18,11,18,25,18,29,18,37,18,25,18,11,18,24,12,47,},
			2,
			7,
			32767,
			65,
			0,
			0,
			0,
			0,
			0,
		};
		check_equals(&tmp, 7);
	}
}

int main(void)
{
	test_save_players();
	return 0;
}
