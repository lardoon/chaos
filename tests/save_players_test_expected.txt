PLAYER:
	NAME "Rich Q"
	COMBAT 1
	RCOMBAT 0
	RANGE 0
	DEFENCE 1
	MOVEMENT 1
	MANOEUVRE 4
	MAGICRES 10
	SPELLC 13
	SPELLS 0,1,23,41,12,42,18,26,18,30,18,3,11,50,18,23,18,45,18,17,18,23,18,37,18,13,
	ABILITY 0
	IMAGE 5
	COLOUR 31775
	TYPE 0
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
PLAYER:
	NAME "Ben"
	COMBAT 7
	RCOMBAT 0
	RANGE 0
	DEFENCE 4
	MOVEMENT 1
	MANOEUVRE 7
	MAGICRES 6
	SPELLC 15
	SPELLS 21,1,18,20,18,5,18,12,18,4,18,9,18,23,18,43,18,29,18,17,18,26,14,56,6,58,6,60,5,59,
	ABILITY 0
	IMAGE 4
	COLOUR 32736
	TYPE 65
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
PLAYER:
	NAME "Saruman"
	COMBAT 6
	RCOMBAT 0
	RANGE 0
	DEFENCE 6
	MOVEMENT 1
	MANOEUVRE 7
	MAGICRES 9
	SPELLC 15
	SPELLS 18,10,18,35,18,64,18,17,18,26,18,20,18,11,18,6,18,9,18,18,18,61,13,1,11,49,6,58,5,59,
	ABILITY 0
	IMAGE 3
	COLOUR 21140
	TYPE 65
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
PLAYER:
	NAME "Kiff"
	COMBAT 7
	RCOMBAT 0
	RANGE 0
	DEFENCE 7
	MOVEMENT 1
	MANOEUVRE 5
	MAGICRES 10
	SPELLC 15
	SPELLS 18,17,18,29,18,6,18,45,18,4,18,22,18,6,18,16,18,21,18,33,18,26,18,36,17,55,14,1,6,58,
	ABILITY 1
	IMAGE 6
	COLOUR 660
	TYPE 65
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
PLAYER:
	NAME "Grotbags"
	COMBAT 5
	RCOMBAT 0
	RANGE 0
	DEFENCE 4
	MOVEMENT 1
	MANOEUVRE 6
	MAGICRES 9
	SPELLC 15
	SPELLS 23,41,18,39,18,14,18,35,18,17,18,33,18,15,18,33,18,6,18,9,18,37,18,3,18,5,18,11,15,1,
	ABILITY 0
	IMAGE 2
	COLOUR 992
	TYPE 65
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
PLAYER:
	NAME "Tim"
	COMBAT 7
	RCOMBAT 0
	RANGE 0
	DEFENCE 4
	MOVEMENT 1
	MANOEUVRE 6
	MAGICRES 6
	SPELLC 18
	SPELLS 18,1,18,28,18,15,18,31,18,3,18,35,18,36,18,6,18,30,18,33,18,19,18,33,18,36,18,23,12,48,12,42,12,42,11,50,
	ABILITY 0
	IMAGE 0
	COLOUR 31744
	TYPE 65
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
PLAYER:
	NAME "Sooty"
	COMBAT 4
	RCOMBAT 0
	RANGE 0
	DEFENCE 5
	MOVEMENT 1
	MANOEUVRE 5
	MAGICRES 6
	SPELLC 16
	SPELLS 18,22,18,20,18,14,18,17,18,35,18,5,18,32,18,39,18,32,18,31,18,61,18,15,18,32,18,10,18,45,14,1,
	ABILITY 0
	IMAGE 1
	COLOUR 31
	TYPE 65
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
PLAYER:
	NAME "H.Potter"
	COMBAT 6
	RCOMBAT 0
	RANGE 0
	DEFENCE 3
	MOVEMENT 1
	MANOEUVRE 6
	MAGICRES 6
	SPELLC 17
	SPELLS 23,40,23,52,21,1,18,62,18,17,18,29,18,34,18,5,18,32,18,11,18,25,18,29,18,37,18,25,18,11,18,24,12,47,
	ABILITY 2
	IMAGE 7
	COLOUR 32767
	TYPE 65
	FLAG 0
	ILLUSION 0
	SELECTED_SPELL 0
	LAST_SPELL 0
	TIMID 0
